File: backend/functions/src/handlers/lahan.ts
import { onCall, HttpsError, CallableRequest } from "firebase-functions/v2/https";
import { getFirestore, FieldValue, DocumentSnapshot } from "firebase-admin/firestore";

const db = () => getFirestore();

// ─── Konstanta validasi ────────────────────────────────────

const JENIS_LAHAN = new Set(["Sawah", "Kebun", "Pekarangan", "Hutan"]);

const SATUAN_LUAS = new Set(["Ha", "m²", "are", "bahu"]);

function assertOwner(doc: DocumentSnapshot, uid: string): void {
  if (!doc.exists) throw new HttpsError("not-found", "Lahan tidak ditemukan");
  if (doc.data()?.uid !== uid) {
    throw new HttpsError("permission-denied", "Tidak punya akses ke lahan ini");
  }
}

function asCoord(v: unknown): number | null {
  const n = Number(v);
  return v !== undefined && v !== null && !isNaN(n) ? n : null;
}

// ─── tambahLahan ───────────────────────────────────────────

export const tambahLahan = onCall(async (request: CallableRequest) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "Login diperlukan");
  }

  const data = request.data as Record<string, unknown>;

  const nama = ((data.nama as string) ?? "").trim();
  const jenisTanaman = ((data.jenis_tanaman as string) ?? "").trim();
  const jenisLahan = ((data.jenis_lahan as string) ?? "").trim();

  if (!nama) throw new HttpsError("invalid-argument", "Nama lahan diperlukan");
  if (!jenisTanaman) throw new HttpsError("invalid-argument", "Jenis tanaman diperlukan");
  if (!jenisLahan) throw new HttpsError("invalid-argument", "Jenis lahan diperlukan");
  if (!JENIS_LAHAN.has(jenisLahan)) {
    throw new HttpsError("invalid-argument", `Jenis lahan tidak valid: ${jenisLahan}. Pilih: ${[...JENIS_LAHAN].join(", ")}`);
  }

  const luas = Number(data.luas ?? 0);
  if (luas < 0) throw new HttpsError("invalid-argument", "Luas tidak boleh negatif");

  const satuanLuas = ((data.satuan_luas as string) ?? "Ha").trim();
  if (!SATUAN_LUAS.has(satuanLuas)) {
    throw new HttpsError("invalid-argument", `Satuan luas tidak valid: ${satuanLuas}`);
  }

  const latitude = asCoord(data.latitude);
  const longitude = asCoord(data.longitude);

  const docRef = db().collection("lahan").doc();
  await docRef.set({
    uid: request.auth.uid,
    nama,
    jenis_tanaman: jenisTanaman,
    jenis_lahan: jenisLahan,
    luas,
    satuan_luas: satuanLuas,
    emoji: (data.emoji as string | null) ?? null,
    latitude,
    longitude,
    catatan: (data.catatan as string | null) ?? null,
    created_at: FieldValue.serverTimestamp(),
    updated_at: FieldValue.serverTimestamp(),
  });

  return { id: docRef.id };
});

// ─── editLahan ─────────────────────────────────────────────

export const editLahan = onCall(async (request: CallableRequest) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "Login diperlukan");
  }

  const lahanId = request.data.id as string;
  if (!lahanId) throw new HttpsError("invalid-argument", "ID lahan diperlukan");

  const docRef = db().collection("lahan").doc(lahanId);
  assertOwner(await docRef.get(), request.auth.uid);

  const ALLOWED = new Set([
    "nama",
    "jenis_tanaman",
    "jenis_lahan",
    "luas",
    "satuan_luas",
    "emoji",
    "latitude",
    "longitude",
    "catatan",
  ]);

  const updates: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(request.data as Record<string, unknown>)) {
    if (ALLOWED.has(k)) updates[k] = v;
  }

  if (
    "jenis_lahan" in updates &&
    updates.jenis_lahan !== null &&
    !JENIS_LAHAN.has(updates.jenis_lahan as string)
  ) {
    throw new HttpsError("invalid-argument", "Jenis lahan tidak valid");
  }

  if (
    "satuan_luas" in updates &&
    updates.satuan_luas !== null &&
    !SATUAN_LUAS.has(updates.satuan_luas as string)
  ) {
    throw new HttpsError("invalid-argument", "Satuan luas tidak valid");
  }

  if ("luas" in updates) {
    const luas = Number(updates.luas);
    if (isNaN(luas) || luas < 0) {
      throw new HttpsError("invalid-argument", "Luas tidak valid");
    }
    updates.luas = luas;
  }

  updates.updated_at = FieldValue.serverTimestamp();
  await docRef.update(updates);

  return { success: true };
});

// ─── hapusLahan ────────────────────────────────────────────

export const hapusLahan = onCall(async (request: CallableRequest) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "Login diperlukan");
  }

  const lahanId = request.data.id as string;
  if (!lahanId) throw new HttpsError("invalid-argument", "ID lahan diperlukan");

  const docRef = db().collection("lahan").doc(lahanId);
  assertOwner(await docRef.get(), request.auth.uid);
  await docRef.delete();

  return { success: true };
});

// ─── getLahanList ──────────────────────────────────────────

export const getLahanList = onCall(async (request: CallableRequest) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "Login diperlukan");
  }

  const snap = await db()
    .collection("lahan")
    .where("uid", "==", request.auth.uid)
    .orderBy("created_at", "desc")
    .get();

  return { lahan: snap.docs.map((d) => ({ id: d.id, ...d.data() })) };
});

// ─── getDetailLahan ────────────────────────────────────────

export const getDetailLahan = onCall(async (request: CallableRequest) => {
  if (!request.auth) {
    throw new HttpsError("unauthenticated", "Login diperlukan");
  }

  const lahanId = request.data.id as string;
  if (!lahanId) throw new HttpsError("invalid-argument", "ID lahan diperlukan");

  const doc = await db().collection("lahan").doc(lahanId).get();
  assertOwner(doc, request.auth.uid);

  const aktivitasSnap = await db()
    .collection("aktivitas")
    .where("lahan_id", "==", lahanId)
    .orderBy("tanggal", "desc")
    .limit(5)
    .get();

  return {
    id: doc.id,
    ...doc.data(),
    aktivitas_terbaru: aktivitasSnap.docs.map((a) => ({ id: a.id, ...a.data() })),
  };
});

File: backend/functions/handlers/lahan.py
from firebase_functions import https_fn
from firebase_admin import firestore
from datetime import datetime, timezone


def _db():
    return firestore.client()


FASE_OPTIONS = {"persiapan", "vegetatif", "generatif", "panen"}


def _assert_owner(doc, uid: str) -> None:
    if not doc.exists:
        raise https_fn.HttpsError("not-found", "Lahan tidak ditemukan")
    if doc.get("uid") != uid:
        raise https_fn.HttpsError("permission-denied", "Tidak punya akses ke lahan ini")


@https_fn.on_call()
def tambah_lahan(req: https_fn.CallableRequest) -> dict:
    if not req.auth:
        raise https_fn.HttpsError("unauthenticated", "Login diperlukan")

    data = req.data
    nama = (data.get("nama") or "").strip()
    jenis_tanaman = (data.get("jenis_tanaman") or "").strip()

    if not nama:
        raise https_fn.HttpsError("invalid-argument", "Nama lahan diperlukan")
    if not jenis_tanaman:
        raise https_fn.HttpsError("invalid-argument", "Jenis tanaman diperlukan")

    fase = data.get("fase", "persiapan")
    if fase not in FASE_OPTIONS:
        raise https_fn.HttpsError("invalid-argument", f"Fase tidak valid: {fase}")

    now = datetime.now(timezone.utc)
    doc_ref = _db().collection("lahan").document()
    doc_ref.set({
        "uid": req.auth.uid,
        "nama": nama,
        "jenis_tanaman": jenis_tanaman,
        "luas_ha": float(data.get("luas_ha", 0)),
        "tanggal_tanam": data.get("tanggal_tanam"),  # ISO date string YYYY-MM-DD
        "fase": fase,
        "catatan": data.get("catatan"),
        "created_at": now,
        "updated_at": now,
    })

    return {"id": doc_ref.id}


@https_fn.on_call()
def edit_lahan(req: https_fn.CallableRequest) -> dict:
    if not req.auth:
        raise https_fn.HttpsError("unauthenticated", "Login diperlukan")

    lahan_id = req.data.get("id")
    if not lahan_id:
        raise https_fn.HttpsError("invalid-argument", "ID lahan diperlukan")

    doc_ref = _db().collection("lahan").document(lahan_id)
    _assert_owner(doc_ref.get(), req.auth.uid)

    allowed = {"nama", "jenis_tanaman", "luas_ha", "tanggal_tanam", "fase", "catatan"}
    updates = {k: v for k, v in req.data.items() if k in allowed}

    if "fase" in updates and updates["fase"] not in FASE_OPTIONS:
        raise https_fn.HttpsError("invalid-argument", "Fase tidak valid")

    updates["updated_at"] = datetime.now(timezone.utc)
    doc_ref.update(updates)

    return {"success": True}


@https_fn.on_call()
def hapus_lahan(req: https_fn.CallableRequest) -> dict:
    if not req.auth:
        raise https_fn.HttpsError("unauthenticated", "Login diperlukan")

    lahan_id = req.data.get("id")
    if not lahan_id:
        raise https_fn.HttpsError("invalid-argument", "ID lahan diperlukan")

    doc_ref = _db().collection("lahan").document(lahan_id)
    _assert_owner(doc_ref.get(), req.auth.uid)
    doc_ref.delete()

    return {"success": True}


@https_fn.on_call()
def get_lahan_list(req: https_fn.CallableRequest) -> dict:
    if not req.auth:
        raise https_fn.HttpsError("unauthenticated", "Login diperlukan")

    docs = (
        _db().collection("lahan")
        .where("uid", "==", req.auth.uid)
        .order_by("created_at", direction="DESCENDING")
        .stream()
    )

    return {"lahan": [{"id": d.id, **d.to_dict()} for d in docs]}


@https_fn.on_call()
def get_detail_lahan(req: https_fn.CallableRequest) -> dict:
    if not req.auth:
        raise https_fn.HttpsError("unauthenticated", "Login diperlukan")

    lahan_id = req.data.get("id")
    if not lahan_id:
        raise https_fn.HttpsError("invalid-argument", "ID lahan diperlukan")

    doc = _db().collection("lahan").document(lahan_id).get()
    _assert_owner(doc, req.auth.uid)

    aktivitas_docs = (
        _db().collection("aktivitas")
        .where("lahan_id", "==", lahan_id)
        .order_by("tanggal", direction="DESCENDING")
        .limit(5)
        .stream()
    )

    return {
        "id": doc.id,
        **doc.to_dict(),
        "aktivitas_terbaru": [{"id": a.id, **a.to_dict()} for a in aktivitas_docs],
    }
